File: python/Balanced_Binary_Tree.py
# Given a binary tree, determine if it is height-balanced.

# For this problem, a height-balanced binary tree is defined as a binary tree in which the depth of the two subtrees of every node never differ by more than 1.


# Definition for a  binary tree node
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    # @param root, a tree node
    # @return a boolean
    def isBalanced(self, root):
        if root is None:
            return True
        if abs(self.depth(root.right) - self.depth(root.left)) > 1:
            return False
        return self.isBalanced(root.left) and self.isBalanced(root.right)

    def depth(self, root):
        if root is None:
            return 0
        leftDepth = self.depth(root.left)
        rightDepth = self.depth(root.right)
        return 1 + max(leftDepth, rightDepth)

if __name__ == "__main__" :
    root = TreeNode(1)
    root.left = TreeNode(1)
    root.right = TreeNode(1)
    s = Solution()
    print s.isBalanced(root)
    print "hello"

File: c++/Balanced_Binary_Tree.cpp
// Given a binary tree, determine if it is height-balanced.

// For this problem, a height-balanced binary tree is defined as a binary tree in which the depth of the two subtrees of every node never differ by more than 1.
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
	bool isBalanced(TreeNode *root) {
	  if (root == NULL){
		return true;
	  }
	  int diff = this -> depth(root -> right) - this -> depth(root -> left);
	  if (diff > 1 || diff < -1){
		return false;
	  }
	  return this -> isBalanced(root -> left) && this -> isBalanced(root -> right);
	};
	int depth(TreeNode *root) {
	  if (root == NULL) {
		return 0;
	  }
	  int leftDepth = this->depth(root -> left);
	  int rightDepth = this->depth(root -> right);
	  if (leftDepth > rightDepth){
		return 1 + leftDepth;
	  }
	  return 1 + rightDepth;
	}
};
